File: go/parsing-log-files/parsing_log_files.go
package parsinglogfiles

import (
	"fmt"
	"regexp"
)

func IsValidLine(text string) bool {
	const validLine string = `^\[(TRC|DBG|INF|WRN|ERR|FTL)\]`

	var validLineRegex = regexp.MustCompile(validLine)

	return validLineRegex.MatchString(text)
}

func SplitLogLine(text string) []string {
	const splitter string = `\<(\~|\*|\=|\-)*\>`

	var splitterRegexp = regexp.MustCompile(splitter)

	return splitterRegexp.Split(text, -1)
}

func CountQuotedPasswords(lines []string) int {
	const quotedPasswords string = `(?i)".*password.*"`

	var quotedPasswordsRegex = regexp.MustCompile(quotedPasswords)

	var count = 0

	for _, line := range lines {
		if quotedPasswordsRegex.MatchString(line) {
			count += 1
		}
	}

	return count
}

func RemoveEndOfLineText(text string) string {
	const endOfLine string = `end-of-line(\d+)`
	const deletedString string = ""

	var endOfLineRegex = regexp.MustCompile(endOfLine)

	return endOfLineRegex.ReplaceAllString(text, deletedString)
}

func TagWithUserName(lines []string) []string {

	const userFinder = `User\s+(\S+)`
	const userPrefixed = "[USR] %s %s"

	var userFinderRegex = regexp.MustCompile(userFinder)

	var newLines []string = make([]string, len(lines))

	for index, line := range lines {

		userMatch := userFinderRegex.FindStringSubmatch(line)
		userExist := userMatch != nil

		if userExist {
			user := userMatch[1]
			newLines[index] = fmt.Sprintf(userPrefixed, user, line)
		} else {
			newLines[index] = line
		}
	}

	return newLines
}

File: csharp/parsing-log-files/ParsingLogFiles.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

public class LogParser
{
    public bool IsValidLine(string text)
    {
        string pattern = @"^\[(TRC|DBG|INF|WRN|ERR|FTL)\]*";
        return new Regex(pattern)
            .Match(text)
            .Success;
    }

    public string[] SplitLogLine(string text)
    {
        string pattern = @"<[\^*=-]+>";
        return Regex.Split(text, pattern);
    }

    public int CountQuotedPasswords(string lines)
    {
        string pattern = @""".*password.*""";
        return Regex
            .Matches(lines, pattern, RegexOptions.IgnoreCase)
            .Count;
    }

    public string RemoveEndOfLineText(string line)
    {
        string pattern = @"end-of-line\d+";
        return new Regex(pattern).Replace(line, "");
    }

    public string[] ListLinesWithPasswords(string[] lines)
        => lines.Select(PasswordProcess).ToArray();

    private string PasswordProcess(string line)
    {
        string pattern = @"password\w+";
        Regex regex = new(pattern, RegexOptions.IgnoreCase);
        Match match = regex.Match(line);

        return $"{(match.Success ? match.Value : "--------")}: {line}";
    }
}
